server,storage: introduce TestCluster

andreimatei opened the pull request:
cc @WillHaack, @RaduBerinde, @dt 

next to come - control and introspection over range leases

fixes #7450

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7448)

<!-- Reviewable:end -->

andreimatei committed:
server: create TestCluster

Takes the code that used to be in multinode_test.go and makes
the TestCluster a proper citizen, in testutils/testcluster.

Also adds capabilities for manually controlling splitting and
replication in TestCluster.

andreimatei committed:
storage: move functions for disabling queues

They'll be joined by other similar Set*QueueActive funs that will be
triggered by StoreTestingKnobs.

RaduBerinde commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 15 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[server/testserver.go, line 311 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLNSip3d2yiVMN0kBxW:-KLNSip3d2yiVMN0kBxX:-423968672) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6bd66060826fc2f2afb0bae9f7017800b15d6a3/server/testserver.go#L311)):*
> ```Go
> func (ts *TestServer) GetFirstStoreID() roachpb.StoreID {
> 	firstStoreID := roachpb.StoreID(-1)
> 	if err := ts.Stores().VisitStores(func(s *storage.Store) error {
> ```

[nit] I would move out the "err := " and put the `if` after, it's pretty hard to read with multiline the function argument

---

*[server/testutils/main_test.go, line 1 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLNSwI_0AcuNhl5-Eod:-KLNSwI_0AcuNhl5-Eoe:676036456) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6bd66060826fc2f2afb0bae9f7017800b15d6a3/server/testutils/main_test.go#L1)):*
> ```Go
> // Copyright 2015 The Cockroach Authors.
> ```

2016

---

*[server/testutils/testcluster.go, line 103 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLNap1d-Z7071rlvitf:-KLNap1eYh58phztx2x2:-339592480) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6bd66060826fc2f2afb0bae9f7017800b15d6a3/server/testutils/testcluster.go#L103)):*
> ```Go
> 		log.Fatalf("cannot split range %+v at start key %q", origRangeDesc, splitKey)
> 	}
> 	splitReq := adminSplitArgs(splitKey)
> ```

this function doesn't help much, things would be more evident if we just initialized the AdminSplitRequest here.

---

*[storage/store.go, line 2498 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLNaDdfcNcVp-O93VTv:-KLNaDdfcNcVp-O93VTw:705557369) ([raw file](https://github.com/cockroachdb/cockroach/blob/e026a7d2355195f2be5c2e5f51693d8fe72eccf1/storage/store.go#L2498)):*
> ```Go
> 
> // DisableReplicateQueue disables or enables the replication queue.
> func (s *Store) DisableReplicateQueue(disabled bool) {
> ```

These function names are pretty unfortunate. Maybe SetEnabledReplicateQueue?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-:-KLNbKgSVbqu6dRY5h68:1417088721)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 2 of 2 files at r1, 4 of 4 files at r2, 11 of 11 files at r3, 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[cli/flags_test.go, line 52 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLXNLYej5lzZYlhI2QD:-KLXNLYfkkjJcwm0eVkP:-593865635) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6bd66060826fc2f2afb0bae9f7017800b15d6a3/cli/flags_test.go#L52)):*
> ```Go
> 			"go/build", // probably not something we want in the main binary
> 			"github.com/cockroachdb/cockroach/security/securitytest", // contains certificates
> 			"github.com/cockroachdb/cockroach/server/testutils",      // meant for testing code only
> ```

What's the comment supposed to tell me?

---

*[kv/dist_sender_server_test.go, line 543 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLXMqeFbnnTQGHhnDwc:-KLXMqeFbnnTQGHhnDwd:-1438638921) ([raw file](https://github.com/cockroachdb/cockroach/blob/684e3411a464cff3072d288ed8090a84b2cdd281/kv/dist_sender_server_test.go#L543)):*
> ```Go
> 	ctx := server.MakeTestContext()
> 	ctx.TestingKnobs.Store = &storeKnobs
> 	s := server.StartTestServerWithContext(t, &ctx, nil /* stopper */)
> ```

Why not just pass stoppers in here?

---

*[server/admin_test.go, line 689 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLXNU7tj9amP45pdesV:-KLXNU7tj9amP45pdesW:-612908725) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6bd66060826fc2f2afb0bae9f7017800b15d6a3/server/admin_test.go#L689)):*
> ```Go
> }
> 
> func TestClusterApi(t *testing.T) {
> ```

s/Api/API/

---

*[server/testserver.go, line 200 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLXN1_tnw666O0m8EjY:-KLXN1_tnw666O0m8EjZ:656992825) ([raw file](https://github.com/cockroachdb/cockroach/blob/684e3411a464cff3072d288ed8090a84b2cdd281/server/testserver.go#L200)):*
> ```Go
> 
> // StartWithStopper is the same as Start, but allows passing a stopper
> // explicitly. Calling stopper.Stop() has the same effect as TestServer.Stop().
> ```

Now that you're passing a stopper, always pass the stopper and remove `.Stop()` - you've made it obsolete and the current contract is a bit messy.

---

*[server/testserver.go, line 265 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLXPXTRnQPjIWmGsxu8:-KLXPXTRnQPjIWmGsxu9:1022072519) ([raw file](https://github.com/cockroachdb/cockroach/blob/e026a7d2355195f2be5c2e5f51693d8fe72eccf1/server/testserver.go#L265)):*
> ```Go
> 		return nil
> 	}
> 	if err := ts.WaitForInitialSplits(); err != nil {
> ```

I think you still want the initial splits to happen or you're going to miss out on some interesting phenomena in tests.

---

*[server/testutils/testcluster.go, line 103 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLNap1d-Z7071rlvitf:-KLXNpRcrH9ApLyosxMz:5172523) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6bd66060826fc2f2afb0bae9f7017800b15d6a3/server/testutils/testcluster.go#L103)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> this function doesn't help much, things would be more evident if we just initialized the AdminSplitRequest here.
</details>
+1

---

*[server/testutils/testcluster.go, line 186 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLXO4w_GuNd4P2TZBe2:-KLXO4w_GuNd4P2TZBe3:-402345927) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6bd66060826fc2f2afb0bae9f7017800b15d6a3/server/testutils/testcluster.go#L186)):*
> ```Go
> 		}
> 	}
> 	panic("unreached")
> ```

This isn't unreachable code, so the message should be better. And in fact, I think at the callsite just above you assume that it returns `nil` here (which might well be better). Mention whatever happens in that case in the comment.

---

*[server/testutils/testcluster_test.go, line 31 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLXOmG4f4qAFmA423q-:-KLXOmG4f4qAFmA423q0:1569126826) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6bd66060826fc2f2afb0bae9f7017800b15d6a3/server/testutils/testcluster_test.go#L31)):*
> ```Go
> )
> 
> func TestTestCluster(t *testing.T) {
> ```

this doesn't really test `TestCluster`.

---

*[server/testutils/testcluster_test.go, line 33 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLXObRb0ivuxY5X46MD:-KLXObRb0ivuxY5X46ME:681113972) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6bd66060826fc2f2afb0bae9f7017800b15d6a3/server/testutils/testcluster_test.go#L33)):*
> ```Go
> func TestTestCluster(t *testing.T) {
> 	defer leaktest.AfterTest(t)()
> 	defer tracing.Disable()()
> ```

Why is that here? Blind copy-pasta or any particular reason?

---

*[storage/client_test.go, line 701 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLXOyC2sT0noUrzNYqV:-KLXOyC2sT0noUrzNYqW:2009668909) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6bd66060826fc2f2afb0bae9f7017800b15d6a3/storage/client_test.go#L701)):*
> ```Go
> 
> 	for _, dest := range dests {
> 		// Perform a consistent read to get the updated range descriptor (as oposed
> ```

s/opo/oppo/

---

*[storage/store.go, line 2498 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLNaDdfcNcVp-O93VTv:-KLXP4oW3dp-L1Ul0Eo-:422076897) ([raw file](https://github.com/cockroachdb/cockroach/blob/e026a7d2355195f2be5c2e5f51693d8fe72eccf1/storage/store.go#L2498)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> These function names are pretty unfortunate. Maybe SetEnabledReplicateQueue?
</details>
or `SetReplicateQueueActive(bool)`

---

*[storage/stores.go, line 248 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLXPE6b_PCNa650mmFa:-KLXPE6b_PCNa650mmFb:-2122253684) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6bd66060826fc2f2afb0bae9f7017800b15d6a3/storage/stores.go#L248)):*
> ```Go
> // RangeLookup implements the RangeDescriptorDB interface. It looks up
> // the descriptors for the given (meta) key.
> // !!! who uses this?
> ```

DistSender in tests. Remove this comment (when you've verified I'm still right).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
the commit message `server, storage: stop replication in TestServer` looks wrong. You're talking about `TestCluster`, no?

bdarnell commented:
Review status: all files reviewed at latest revision, 25 unresolved discussions, some commit checks failed.

---

*[cli/flags_test.go, line 52 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLXNLYej5lzZYlhI2QD:-KLcDkiBNDKBKhMFGU33:-979821350) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6bd66060826fc2f2afb0bae9f7017800b15d6a3/cli/flags_test.go#L52)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> What's the comment supposed to tell me?
</details>
Same thing as the one two lines down. We generally put testing utilities in `testutils/fooutils` instead of `foo/testutils`

---

*[server/testutils/main_test.go, line 24 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLcEkYeqExJjAa8RR-q:-KLcEkYf1GDAKgZqb-GK:-2090825410) ([raw file](https://github.com/cockroachdb/cockroach/blob/e026a7d2355195f2be5c2e5f51693d8fe72eccf1/server/testutils/main_test.go#L24)):*
> ```Go
> )
> 
> func init() {
> ```

Also add `leaktest`

---

*[server/testutils/testcluster.go, line 46 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLcGVm7aCslRmRATt8u:-KLcGVm7aCslRmRATt8v:87013123) ([raw file](https://github.com/cockroachdb/cockroach/blob/e026a7d2355195f2be5c2e5f51693d8fe72eccf1/server/testutils/testcluster.go#L46)):*
> ```Go
> }
> 
> // lookupRange returns the descriptor of the range containing key.
> ```

Why do you need to go into implementation details like this instead of sending a LookupReplica RPC?

---

*[server/testutils/testcluster.go, line 95 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLcGrX_wUbTK9LQt1YC:-KLcGrX_wUbTK9LQt1YD:-2016928861) ([raw file](https://github.com/cockroachdb/cockroach/blob/e026a7d2355195f2be5c2e5f51693d8fe72eccf1/server/testutils/testcluster.go#L95)):*
> ```Go
> // The right range created by the split starts at the split key and extends to the
> // original range's end key.
> // Returns the new ids of the left and right ranges.
> ```

s/ids/descriptors/

---

*[server/testutils/testcluster.go, line 99 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLcH53XBERNbKmVT1h8:-KLcH53XBERNbKmVT1h9:1456439701) ([raw file](https://github.com/cockroachdb/cockroach/blob/e026a7d2355195f2be5c2e5f51693d8fe72eccf1/server/testutils/testcluster.go#L99)):*
> ```Go
> 	splitKey roachpb.Key,
> ) (*roachpb.RangeDescriptor, *roachpb.RangeDescriptor) {
> 	origRangeDesc := tc.lookupRange(addr(splitKey))
> ```

Assign `addr(splitKey)` to a local variable instead of addressing it three separate times. (Then maybe you don't need the helper function)

---

*[server/testutils/testcluster.go, line 101 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLcHHhSeBevzgBTWXl2:-KLcHHhSeBevzgBTWXl3:-1480499585) ([raw file](https://github.com/cockroachdb/cockroach/blob/e026a7d2355195f2be5c2e5f51693d8fe72eccf1/server/testutils/testcluster.go#L101)):*
> ```Go
> 	origRangeDesc := tc.lookupRange(addr(splitKey))
> 	if origRangeDesc.StartKey.Equal(addr(splitKey)) {
> 		log.Fatalf("cannot split range %+v at start key %q", origRangeDesc, splitKey)
> ```

Avoid `log.Fatalf` and `panic` in tests. Either pass a `testing.T` around and use `T.Fatalf` or make this method return an error.

---

*[server/testutils/testcluster.go, line 109 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLcHi0PIiimMibHCGvd:-KLcHi0PIiimMibHCGve:-55691798) ([raw file](https://github.com/cockroachdb/cockroach/blob/e026a7d2355195f2be5c2e5f51693d8fe72eccf1/server/testutils/testcluster.go#L109)):*
> ```Go
> 	}
> 	leftRangeDesc := tc.lookupRange(origRangeDesc.StartKey)
> 	rightRangeDesc := tc.lookupRange(addr(splitKey))
> ```

This is racy: at this point we know that at least one store has processed the split, but `tc.lookupRange` could find an older pre-split descriptor.

---

*[server/testutils/testcluster.go, line 141 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLcKc_TKPgCHO3f1_eL:-KLcKc_TKPgCHO3f1_eM:1722625637) ([raw file](https://github.com/cockroachdb/cockroach/blob/e026a7d2355195f2be5c2e5f51693d8fe72eccf1/server/testutils/testcluster.go#L141)):*
> ```Go
> 		}
> 
> 		// Ask a random replica of the range to up-replicate.
> ```

At this level you should probably just send a ChangeReplicas RPC instead of accessing a Replica object directly.

---

*[server/testutils/testcluster.go, line 191 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLcJuBvPj6_EqxTtkz9:-KLcJuBw2-DAvE78P1sy:1898047018) ([raw file](https://github.com/cockroachdb/cockroach/blob/e026a7d2355195f2be5c2e5f51693d8fe72eccf1/server/testutils/testcluster.go#L191)):*
> ```Go
> // StartTestCluster starts up a TestCluster made up of `nodes` in-memory testing
> // servers, and creates database `name`.
> // The cluster has replication disabled; all replication must be controlled
> ```

`TestTestCluster` disables the replication queue separately. Is this a stale comment?

---

*[server/testutils/testcluster.go, line 193 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLcJUUobl-eTksxKFSu:-KLcJUUobl-eTksxKFSv:1314942124) ([raw file](https://github.com/cockroachdb/cockroach/blob/e026a7d2355195f2be5c2e5f51693d8fe72eccf1/server/testutils/testcluster.go#L193)):*
> ```Go
> // The cluster has replication disabled; all replication must be controlled
> // manually.
> // Also returns a cleanup func that stops and cleans up all nodes and
> ```

Why a separate func instead of a `Stop()` method on the `TestCluster`?

---

*[server/testutils/testcluster_test.go, line 66 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLcFQ9dfetirPs4EKdt:-KLcFQ9dfetirPs4EKdu:907611473) ([raw file](https://github.com/cockroachdb/cockroach/blob/e026a7d2355195f2be5c2e5f51693d8fe72eccf1/server/testutils/testcluster_test.go#L66)):*
> ```Go
> 	}
> 
> 	// Split the table to a new range.
> ```

Comment that this would normally happen automatically, but we've disabled the split queue in this test. 

---

*[server/testutils/testcluster_test.go, line 73 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLcFFb1NSiZekwn7_vB:-KLcFFb1NSiZekwn7_vC:-74715257) ([raw file](https://github.com/cockroachdb/cockroach/blob/e026a7d2355195f2be5c2e5f51693d8fe72eccf1/server/testutils/testcluster_test.go#L73)):*
> ```Go
> 	leftRangeDesc, tableRangeDesc := tc.SplitRange(tableStartKey)
> 	log.Infof("After split got ranges: %+v and %+v.", leftRangeDesc, tableRangeDesc)
> 	if tableRangeDesc.Replicas[0].NodeID != 1 {
> ```

Looks like you want to check `len(tableRangeDesc.Replicas)` too.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448)*
<!-- Sent from Reviewable.io -->

andreimatei committed:
storage: make store.LookupReplica() public

To be used by the TestCluster when controlling replication.

andreimatei commented:
Review status: 0 of 19 files reviewed at latest revision, 25 unresolved discussions, some commit checks pending.

---

*[server/testserver.go, line 200 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLXN1_tnw666O0m8EjY:-KMQhOT6_xmjlxVFelVx:-1558106152) ([raw file](https://github.com/cockroachdb/cockroach/blob/684e3411a464cff3072d288ed8090a84b2cdd281/server/testserver.go#L200)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Now that you're passing a stopper, always pass the stopper and remove `.Stop()` - you've made it obsolete and the current contract is a bit messy.
</details>
this is gone

---

*[server/testserver.go, line 311 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLNSip3d2yiVMN0kBxW:-KMQmWfUpKQEdaI_1Br9:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6bd66060826fc2f2afb0bae9f7017800b15d6a3/server/testserver.go#L311)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> [nit] I would move out the "err := " and put the `if` after, it's pretty hard to read with multiline the function argument
</details>
Done.

---

*[server/testserver.go, line 265 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLXPXTRnQPjIWmGsxu8:-KMQlmrBBqEIlEuYlmRl:882532292) ([raw file](https://github.com/cockroachdb/cockroach/blob/e026a7d2355195f2be5c2e5f51693d8fe72eccf1/server/testserver.go#L265)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> I think you still want the initial splits to happen or you're going to miss out on some interesting phenomena in tests.
</details>
well the interface we have now I think only works for disabling the split queue from the very beginning, when the store is created.

---

*[storage/client_test.go, line 701 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLXOyC2sT0noUrzNYqV:-KMQmbpL-rdjG-yabM1F:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6bd66060826fc2f2afb0bae9f7017800b15d6a3/storage/client_test.go#L701)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> s/opo/oppo/
</details>
Done.

---

*[storage/store.go, line 2498 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLNaDdfcNcVp-O93VTv:-KLb5BR_Fz4pzZfmpLOy:1211166229) ([raw file](https://github.com/cockroachdb/cockroach/blob/e026a7d2355195f2be5c2e5f51693d8fe72eccf1/storage/store.go#L2498)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> or `SetReplicateQueueActive(bool)`
</details>
went with `Set*QueueActive(active bool)`

---

*[storage/stores.go, line 248 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLXPE6b_PCNa650mmFa:-KLbFRyepp3fiforSgnM:84912821) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6bd66060826fc2f2afb0bae9f7017800b15d6a3/storage/stores.go#L248)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> DistSender in tests. Remove this comment (when you've verified I'm still right).
</details>
This guy seems to be used by `LocalTestCluster`, which seems to be some ancient test contraption.

---

*[cli/flags_test.go, line 52 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLXNLYej5lzZYlhI2QD:-KLYRRETDdlGRVIWjNVi:185863926) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6bd66060826fc2f2afb0bae9f7017800b15d6a3/cli/flags_test.go#L52)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> What's the comment supposed to tell me?
</details>
as below, that the testutils package is meant just for tests

---

*[kv/dist_sender_server_test.go, line 543 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLXMqeFbnnTQGHhnDwc:-KLbXbC1aNZspSopj7CA:-1780569939) ([raw file](https://github.com/cockroachdb/cockroach/blob/684e3411a464cff3072d288ed8090a84b2cdd281/kv/dist_sender_server_test.go#L543)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Why not just pass stoppers in here?
</details>
ha, what "stoppers"?
Do you mean that the stopper should not be an optional argument? And that there should be no `TestServer.Stop()` method, but have the client call `stopper.Stop()`? I've decided to not go that far with this refactoring...
I think when I integrate with Will's PR, this `stopper` arg will be hidden inside a `Params` struct.

---

*[server/admin_test.go, line 689 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLXNU7tj9amP45pdesV:-KLbYM9UeoD-e_VS6Evp:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6bd66060826fc2f2afb0bae9f7017800b15d6a3/server/admin_test.go#L689)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> s/Api/API/
</details>
Done.

---

*[server/testutils/main_test.go, line 1 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLNSwI_0AcuNhl5-Eod:-KMBD7AEpqlq0Huzic1I:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6bd66060826fc2f2afb0bae9f7017800b15d6a3/server/testutils/main_test.go#L1)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> 2016
</details>
Done.

---

*[server/testutils/main_test.go, line 24 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLcEkYeqExJjAa8RR-q:-KMBDa-U_cFfDhg5si5x:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e026a7d2355195f2be5c2e5f51693d8fe72eccf1/server/testutils/main_test.go#L24)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Also add `leaktest`
</details>
Done.

---

*[server/testutils/testcluster.go, line 103 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLNap1d-Z7071rlvitf:-KMQn8TDqmV8KTWsTLfm:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6bd66060826fc2f2afb0bae9f7017800b15d6a3/server/testutils/testcluster.go#L103)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> +1
</details>
Done.

---

*[server/testutils/testcluster.go, line 186 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLXO4w_GuNd4P2TZBe2:-KMROpwMJRlsTklHYM95:-1127821891) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6bd66060826fc2f2afb0bae9f7017800b15d6a3/server/testutils/testcluster.go#L186)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> This isn't unreachable code, so the message should be better. And in fact, I think at the callsite just above you assume that it returns `nil` here (which might well be better). Mention whatever happens in that case in the comment.
</details>
made the function return an error, but the error is always unexpected. If you have a StoreID, the store really should exist somewhere.

---

*[server/testutils/testcluster.go, line 46 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLcGVm7aCslRmRATt8u:-KMR_w2YxoqSF3sIXuxm:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e026a7d2355195f2be5c2e5f51693d8fe72eccf1/server/testutils/testcluster.go#L46)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Why do you need to go into implementation details like this instead of sending a LookupReplica RPC?
</details>
Done.

---

*[server/testutils/testcluster.go, line 95 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLcGrX_wUbTK9LQt1YC:-KMQnFs3bdGk5UAUKjjA:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e026a7d2355195f2be5c2e5f51693d8fe72eccf1/server/testutils/testcluster.go#L95)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> s/ids/descriptors/
</details>
Done.

---

*[server/testutils/testcluster.go, line 99 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLcH53XBERNbKmVT1h8:-KMQntynzVJGnOChmnN0:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e026a7d2355195f2be5c2e5f51693d8fe72eccf1/server/testutils/testcluster.go#L99)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Assign `addr(splitKey)` to a local variable instead of addressing it three separate times. (Then maybe you don't need the helper function)
</details>
Done.

---

*[server/testutils/testcluster.go, line 101 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLcHHhSeBevzgBTWXl2:-KMQpVy1QqQZeQr60m11:2066118171) ([raw file](https://github.com/cockroachdb/cockroach/blob/e026a7d2355195f2be5c2e5f51693d8fe72eccf1/server/testutils/testcluster.go#L101)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Avoid `log.Fatalf` and `panic` in tests. Either pass a `testing.T` around and use `T.Fatalf` or make this method return an error.
</details>
ok, changed mosts. The reason why I've avoided passing a testing.T for most of these methods is that I expect them to be called from other goroutines than the test's main one, so error handling is more difficult. But in the meantime I've shifted my view some towards trying to avoid panics in these cases too.

---

*[server/testutils/testcluster.go, line 109 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLcHi0PIiimMibHCGvd:-KMRJrM8ybqDwh5TEd9i:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e026a7d2355195f2be5c2e5f51693d8fe72eccf1/server/testutils/testcluster.go#L109)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> This is racy: at this point we know that at least one store has processed the split, but `tc.lookupRange` could find an older pre-split descriptor.
</details>
Done.

---

*[server/testutils/testcluster.go, line 141 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLcKc_TKPgCHO3f1_eL:-KMRNB2jFpv6wRrDs2jx:-1269795519) ([raw file](https://github.com/cockroachdb/cockroach/blob/e026a7d2355195f2be5c2e5f51693d8fe72eccf1/server/testutils/testcluster.go#L141)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> At this level you should probably just send a ChangeReplicas RPC instead of accessing a Replica object directly.
</details>
Is there such an RPC?

---

*[server/testutils/testcluster.go, line 191 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLcJuBvPj6_EqxTtkz9:-KMRK5TsHtpvnPdSHD_d:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e026a7d2355195f2be5c2e5f51693d8fe72eccf1/server/testutils/testcluster.go#L191)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> `TestTestCluster` disables the replication queue separately. Is this a stale comment?
</details>
Done.

---

*[server/testutils/testcluster.go, line 193 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLcJUUobl-eTksxKFSu:-KMQqHeG00EKba2BWg6q:677677860) ([raw file](https://github.com/cockroachdb/cockroach/blob/e026a7d2355195f2be5c2e5f51693d8fe72eccf1/server/testutils/testcluster.go#L193)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Why a separate func instead of a `Stop()` method on the `TestCluster`?
</details>
gone

---

*[server/testutils/testcluster_test.go, line 31 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLXOmG4f4qAFmA423q-:-KMBDqIlaxhf4MWnXLZx:248552160) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6bd66060826fc2f2afb0bae9f7017800b15d6a3/server/testutils/testcluster_test.go#L31)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> this doesn't really test `TestCluster`.
</details>
it does... It exercises it some :)
Anyway, now changed the name a bit, maybe that'll make you like it more.

---

*[server/testutils/testcluster_test.go, line 33 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLXObRb0ivuxY5X46MD:-KMBDdWjuAkcD7FQH0rj:609326913) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6bd66060826fc2f2afb0bae9f7017800b15d6a3/server/testutils/testcluster_test.go#L33)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Why is that here? Blind copy-pasta or any particular reason?
</details>
gone. copy-pasta.

---

*[server/testutils/testcluster_test.go, line 66 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLcFQ9dfetirPs4EKdt:-KMRXVRjHbCv6paKrd1F:850919778) ([raw file](https://github.com/cockroachdb/cockroach/blob/e026a7d2355195f2be5c2e5f51693d8fe72eccf1/server/testutils/testcluster_test.go#L66)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Comment that this would normally happen automatically, but we've disabled the split queue in this test. 
</details>
this test is now called `TestTestClusterWithManualReplication`, so I think we're probably good. Also I'm not sure exactly what you were trying to emphasize? Does each table get its own range? Or the first user table?

---

*[server/testutils/testcluster_test.go, line 73 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLcFFb1NSiZekwn7_vB:-KMRYBDet4BZk9QhUHaQ:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e026a7d2355195f2be5c2e5f51693d8fe72eccf1/server/testutils/testcluster_test.go#L73)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Looks like you want to check `len(tableRangeDesc.Replicas)` too.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 15 of 15 files at r5, 9 of 9 files at r6, 6 of 8 files at r7.
Review status: 17 of 19 files reviewed at latest revision, 30 unresolved discussions, some commit checks failed.

---

*[server/testserver.go, line 265 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLXPXTRnQPjIWmGsxu8:-KMUS8zwqHPeGNQqp0P8:-65167494) ([raw file](https://github.com/cockroachdb/cockroach/blob/e026a7d2355195f2be5c2e5f51693d8fe72eccf1/server/testserver.go#L265)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> well the interface we have now I think only works for disabling the split queue from the very beginning, when the store is created.
</details>
Seems odd to keep `config.TestingTableSplitsDisabled` here while introducing `stk.DisableSplitQueue`.

---

*[server/testserver.go, line 404 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KMUSROoYE8EU1GmJ3e1:-KMUSROoYE8EU1GmJ3e2:-372919782) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4e7deb279dfd0589ed97c63b1f8e39fd6d946b0/server/testserver.go#L404)):*
> ```Go
> 	})
> 	if err != nil {
> 		panic(err)
> ```

can you please return an error?

---

*[sql/bench_test.go, line 56 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KMURpVwzymtMypVzkpe:-KMURpVwzymtMypVzkpf:1866978281) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4e7deb279dfd0589ed97c63b1f8e39fd6d946b0/sql/bench_test.go#L56)):*
> ```Go
> 			ReplicationMode: testcluster.ReplicationFull,
> 			ServerArgs: base.TestServerArgs{
> 				UseDatabase: "bench",
> ```

why does `benchmarkMultinodeCockroach` need to know about this database? seems that `f` may use any database it wants.

---

*[storage/replica_command.go, line 2152 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KMURZdBd7egwUE4TSAw:-KMURZdC9PE_70C6AUPb:166874) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4e7deb279dfd0589ed97c63b1f8e39fd6d946b0/storage/replica_command.go#L2152)):*
> ```Go
> 		if err != nil {
> 			return reply, roachpb.NewErrorf("cannot split range at key %s: %v",
> 				foundSplitKey, err)
> ```

?

---

*[storage/store.go, line 2577 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KMURHfaQIL4l_sIWU-d:-KMURHfaQIL4l_sIWU-e:1907232165) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4e7deb279dfd0589ed97c63b1f8e39fd6d946b0/storage/store.go#L2577)):*
> ```Go
> 
> // SetReplicaGCQueueActive enables or disables the replica GC queue.
> func (s *Store) SetReplicaGCQueueActive(active bool) {
> ```

since these functions are controlled via testing knobs, why do they need to be exported? Is this just a halfway refactoring?

---

*[testutils/testcluster/testcluster.go, line 79 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KMUVva5uEnhxQb18EQR:-KMUVva5uEnhxQb18EQS:403410462) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4e7deb279dfd0589ed97c63b1f8e39fd6d946b0/testutils/testcluster/testcluster.go#L79)):*
> ```Go
> // servers.
> // The cluster should be stopped using cluster.Stopper().Stop().
> func StartTestCluster(t testing.TB, nodes int, args ClusterArgs) *TestCluster {
> ```

why a pointer?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Thanks for the review y'all! PTAL. `r5`-`r7`.

---

Review status: 17 of 19 files reviewed at latest revision, 30 unresolved discussions, some commit checks failed.

---

*[server/testserver.go, line 265 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLXPXTRnQPjIWmGsxu8:-KMVt6Y8Mod7ZiRM9Bnm:-1270922058) ([raw file](https://github.com/cockroachdb/cockroach/blob/e026a7d2355195f2be5c2e5f51693d8fe72eccf1/server/testserver.go#L265)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Seems odd to keep `config.TestingTableSplitsDisabled` here while introducing `stk.DisableSplitQueue`.
</details>
I've created another PR to get rid of `config.TestingTableSplitsDisabled`
https://github.com/cockroachdb/cockroach/pull/7789

---

*[server/testserver.go, line 404 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KMUSROoYE8EU1GmJ3e1:-KMUcbMyHVlcgKMZvcE7:-313355087) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4e7deb279dfd0589ed97c63b1f8e39fd6d946b0/server/testserver.go#L404)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> can you please return an error?
</details>
this one seems like a clear-cut assertion; if the visitor doesn't return errors, neither does `VisitStores`. No reason for `GetFirstStoreID` to return an error.

---

*[sql/bench_test.go, line 56 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KMURpVwzymtMypVzkpe:-KMUdBqVCYdR5vTprsrD:773927857) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4e7deb279dfd0589ed97c63b1f8e39fd6d946b0/sql/bench_test.go#L56)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why does `benchmarkMultinodeCockroach` need to know about this database? seems that `f` may use any database it wants.
</details>
I just moved the database creation out of the old `SetupMultinodeTestCluster` function

---

*[storage/replica_command.go, line 2152 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KMURZdBd7egwUE4TSAw:-KMUdXPu1UIPGtyuFvJM:256423425) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4e7deb279dfd0589ed97c63b1f8e39fd6d946b0/storage/replica_command.go#L2152)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> ?
</details>
random fix (`splitKey` is only assigned below). Didn't particularly belong in any other commit and I didn't want to do one just for it...

---

*[storage/store.go, line 2577 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KMURHfaQIL4l_sIWU-d:-KMUgV87FGWaUjpsrMG2:559288914) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4e7deb279dfd0589ed97c63b1f8e39fd6d946b0/storage/store.go#L2577)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> since these functions are controlled via testing knobs, why do they need to be exported? Is this just a halfway refactoring?
</details>
Well only the initial state of (some of) the queues is controlled via knobs; the knobs so far only act as input args and callbacks, not as controllers. Later on they can be started/stopped at will. In particular, some tests also use them through the `multiTestContext` (which is in `storage_test`). So I think they'll live to die another day.

---

*[testutils/testcluster/testcluster.go, line 79 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KMUVva5uEnhxQb18EQR:-KMUdpaVhhd_Iktt9vKO:1161720248) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4e7deb279dfd0589ed97c63b1f8e39fd6d946b0/testutils/testcluster/testcluster.go#L79)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why a pointer?
</details>
I think `TestCluster` should be passed around by pointer, like `TestServer`; it doesn't seem like a "value type". It's probably going to get more state in the future.

---

*[cli/flags_test.go, line 52 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KLXNLYej5lzZYlhI2QD:-KMRdapFE0uU4CVyOQpa:868425799) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6bd66060826fc2f2afb0bae9f7017800b15d6a3/cli/flags_test.go#L52)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> as below, that the testutils package is meant just for tests
</details>
`server/testutils` is gone

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-:-KMRdGbqKljtrfUxjCpU:-108131706)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 17 of 19 files reviewed at latest revision, 29 unresolved discussions, some commit checks pending.

---

*[server/testserver.go, line 404 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KMUSROoYE8EU1GmJ3e1:-KMVvwSrJWNXwS_dtEg3:-2035849960) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4e7deb279dfd0589ed97c63b1f8e39fd6d946b0/server/testserver.go#L404)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> this one seems like a clear-cut assertion; if the visitor doesn't return errors, neither does `VisitStores`. No reason for `GetFirstStoreID` to return an error.
</details>
It doesn't matter if it's an assertion or not - what matters is the ergonomics. If this assertion fires, you're going to tank my test suite, and that's not nice.

---

*[storage/replica_command.go, line 2152 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KMURZdBd7egwUE4TSAw:-KMVwTXjtidbfrF7MNrz:483777287) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4e7deb279dfd0589ed97c63b1f8e39fd6d946b0/storage/replica_command.go#L2152)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> random fix (`splitKey` is only assigned below). Didn't particularly belong in any other commit and I didn't want to do one just for it...
</details>
I think this really wants to be `args.SplitKey`, though.

---

*[storage/store.go, line 2577 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KMURHfaQIL4l_sIWU-d:-KMVweeTBAQtRlvaSmgH:-891853881) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4e7deb279dfd0589ed97c63b1f8e39fd6d946b0/storage/store.go#L2577)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> Well only the initial state of (some of) the queues is controlled via knobs; the knobs so far only act as input args and callbacks, not as controllers. Later on they can be started/stopped at will. In particular, some tests also use them through the `multiTestContext` (which is in `storage_test`). So I think they'll live to die another day.
</details>
I don't understand what you're saying. These are used by multiTestContext, so they can't be removed? Can those tests use the knobs?

If yes, can you sprinkle some TODOs?

---

*[testutils/testcluster/testcluster.go, line 79 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KMUVva5uEnhxQb18EQR:-KMVvhJAnYy_Kg2Vtv6a:1217928756) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4e7deb279dfd0589ed97c63b1f8e39fd6d946b0/testutils/testcluster/testcluster.go#L79)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> I think `TestCluster` should be passed around by pointer, like `TestServer`; it doesn't seem like a "value type". It's probably going to get more state in the future.
</details>
It is certainly a value type currently; passing it around as a value is more restrictive (in other words, changing to a pointer later is easy, but going the other way is not), so I think we should do that for now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 17 of 19 files reviewed at latest revision, 29 unresolved discussions, some commit checks failed.

---

*[server/testserver.go, line 404 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KMUSROoYE8EU1GmJ3e1:-KMVxe70NHn_JFETOuiy:-976013163) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4e7deb279dfd0589ed97c63b1f8e39fd6d946b0/server/testserver.go#L404)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> It doesn't matter if it's an assertion or not - what matters is the ergonomics. If this assertion fires, you're going to tank my test suite, and that's not nice.
</details>
Testing code should use assertions like this to the same extent that non-testing code does (e.g. this could be re-written as `Stores` providing a `VisitStores` variant that takes a function that cannot return an error, and internally calling the other version and doing the assertion.
So you're saying that we shouldn't be using panics anywhere in our codebase. That's not what we've been doing. And I think it's going too far. If we ever come to that agreement, it should be in the style guide.

---

*[storage/replica_command.go, line 2152 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KMURZdBd7egwUE4TSAw:-KMVznN5wrD7GR-hLRTh:-1773579464) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4e7deb279dfd0589ed97c63b1f8e39fd6d946b0/storage/replica_command.go#L2152)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I think this really wants to be `args.SplitKey`, though.
</details>
yeah, that's right, thanks

---

*[storage/store.go, line 2577 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KMURHfaQIL4l_sIWU-d:-KMW-Dgwli7-gGB4mTo6:-1534705908) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4e7deb279dfd0589ed97c63b1f8e39fd6d946b0/storage/store.go#L2577)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I don't understand what you're saying. These are used by multiTestContext, so they can't be removed? Can those tests use the knobs?
> 
> If yes, can you sprinkle some TODOs?
</details>
the `mtc` tests enable and disable these queues at several points, as opposed to disabling them from the beginning (which is the only thing you can do through the knobs)

---

*[testutils/testcluster/testcluster.go, line 79 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KMUVva5uEnhxQb18EQR:-KMW-aLmATapEYTXV-EH:400632411) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4e7deb279dfd0589ed97c63b1f8e39fd6d946b0/testutils/testcluster/testcluster.go#L79)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> It is certainly a value type currently; passing it around as a value is more restrictive (in other words, changing to a pointer later is easy, but going the other way is not), so I think we should do that for now.
</details>
Then we'd have to rename it to `TestClusterHandle` or something, to not give the impression that you're actually creating a copy of a cluster.
I don't like that very much. I think the pointer is fine - it says one cluster will be shared.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 8 of 9 files at r9, 7 of 9 files at r10.
Review status: 18 of 20 files reviewed at latest revision, 27 unresolved discussions, some commit checks failed.

---

*[storage/store.go, line 2577 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KMURHfaQIL4l_sIWU-d:-KMWAAEq0IHqInPb_vpE:-1238307696) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4e7deb279dfd0589ed97c63b1f8e39fd6d946b0/storage/store.go#L2577)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> the `mtc` tests enable and disable these queues at several points, as opposed to disabling them from the beginning (which is the only thing you can do through the knobs)
</details>
I see. I guess I'd leave these unexported, then, and only export them in helpers_test.

---

*[testutils/testcluster/testcluster.go, line 79 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KMUVva5uEnhxQb18EQR:-KMWAJDqy6FRqw3V4XTN:1855096922) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4e7deb279dfd0589ed97c63b1f8e39fd6d946b0/testutils/testcluster/testcluster.go#L79)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> Then we'd have to rename it to `TestClusterHandle` or something, to not give the impression that you're actually creating a copy of a cluster.
> I don't like that very much. I think the pointer is fine - it says one cluster will be shared.
</details>
By the same token, dereferencing this should imply making a new cluster, which is not the case.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 15 files at r5, 8 of 9 files at r9, 9 of 9 files at r10.
Review status: all files reviewed at latest revision, 29 unresolved discussions, some commit checks failed.

---

*[server/testserver.go, line 384 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KM_fyQaHjsWBc4poo7n:-KM_fyQb9pulOOGfdYlv:232821600) ([raw file](https://github.com/cockroachdb/cockroach/blob/d33285fa6fcc1eb26d17a891bed4a8c07386940a/server/testserver.go#L384)):*
> ```Go
> 
> // GetNode exposes the Server's Node.
> func (ts *TestServer) GetNode() *Node {
> ```

In general I'd prefer to just export `ts.node` instead of adding getter methods like this one and the next. It's true that a method can be read-only while an exported field can be set, but since you're returning a pointer you're already trusting the caller not to mutate things in appropriately.

---

*[testutils/testcluster/testcluster.go, line 104 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KM_hxcFXGjSQ7dEystC:-KM_hxcFXGjSQ7dEystD:1924194891) ([raw file](https://github.com/cockroachdb/cockroach/blob/d33285fa6fcc1eb26d17a891bed4a8c07386940a/testutils/testcluster/testcluster.go#L104)):*
> ```Go
> 	switch args.ReplicationMode {
> 	case ReplicationFull:
> 		// Force all ranges to be replicated everywhere. This is needed until #7297 is
> ```

#7297 is now closed.

---

*[testutils/testcluster/testcluster.go, line 127 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KM_irdHDtSZZhfxY-RV:-KM_irdHDtSZZhfxY-RW:605515769) ([raw file](https://github.com/cockroachdb/cockroach/blob/d33285fa6fcc1eb26d17a891bed4a8c07386940a/testutils/testcluster/testcluster.go#L127)):*
> ```Go
> 	servers = append(servers, first.(*server.TestServer))
> 	conns = append(conns, conn)
> 	args.ServerArgs.JoinAddr = first.ServingAddr()
> ```

Make a copy of `args.ServerArgs` instead of mutating the value that was passed in, so that the same ClusterArgs can be used more than once. (this is true for all of the mutations you make to args.ServerArgs, but especially this one since it is the one that varies per server)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-:-KM_kS-P5wTzAJZ2xEyl:1417088721)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 2 of 20 files reviewed at latest revision, 29 unresolved discussions, some commit checks pending.

---

*[server/testserver.go, line 384 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KM_fyQaHjsWBc4poo7n:-KMe9_ylTuvMkq3oKwUL:1107173266) ([raw file](https://github.com/cockroachdb/cockroach/blob/d33285fa6fcc1eb26d17a891bed4a8c07386940a/server/testserver.go#L384)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> In general I'd prefer to just export `ts.node` instead of adding getter methods like this one and the next. It's true that a method can be read-only while an exported field can be set, but since you're returning a pointer you're already trusting the caller not to mutate things in appropriately.
</details>
What's the disadvantage of accessors like this? Making a field (not even a method) of a struct as big and important as `Server` public seems like a big hammer. People are gonna start using them. With accessors like this, at least public access is contained to tests and we can trace what tests are doing involved things when we need to refactor.
Maybe an alternative you'd find more palatable would be to move these accessors into a struct under `Server`, e.g. `server.TestAccess`? I've opted against that initially because I wasn't convinced that giving access like this to tests is even the right thing to do, maybe we'll come up with something more targeted for `TestCluster`'s usage in the future.

---

*[storage/store.go, line 2577 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KMURHfaQIL4l_sIWU-d:-KMeAy2JWHLCiUa7rBDj:818278950) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4e7deb279dfd0589ed97c63b1f8e39fd6d946b0/storage/store.go#L2577)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I see. I guess I'd leave these unexported, then, and only export them in helpers_test.
</details>
well the methods are just for tests in the first place, so I don't really see a benefit in having a public version in `helpers_test`. I guess we could append a `ForTest` to all of their names, but I don't think it's worth it.
I'll add a communal comment before their definition.

---

*[testutils/testcluster/testcluster.go, line 79 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KMUVva5uEnhxQb18EQR:-KMeBT1LVBSNXS23xdt1:218675853) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4e7deb279dfd0589ed97c63b1f8e39fd6d946b0/testutils/testcluster/testcluster.go#L79)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> By the same token, dereferencing this should imply making a new cluster, which is not the case.
</details>
let's leave it, on the account that we might get adding/removing servers or other state-changing operations in the future

---

*[testutils/testcluster/testcluster.go, line 104 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KM_hxcFXGjSQ7dEystC:-KMe88yxVCArwphvwqOf:1890780918) ([raw file](https://github.com/cockroachdb/cockroach/blob/d33285fa6fcc1eb26d17a891bed4a8c07386940a/testutils/testcluster/testcluster.go#L104)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> #7297 is now closed.
</details>
yeah... So this means that most tests should not want "full replication", but just default 3x replication. But I'm not sure what the best way of rewriting `WaitForFullReplication()` to block for that (blocking for "full replication" seems easy and brutal.
I've added a TODO to remove the `ReplicationFull` mode, or at least not have it be the default.
Maybe @willhaack will have ideas.

---

*[testutils/testcluster/testcluster.go, line 127 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KM_irdHDtSZZhfxY-RV:-KMe7PkUldMuhJMuJOqE:-1265599335) ([raw file](https://github.com/cockroachdb/cockroach/blob/d33285fa6fcc1eb26d17a891bed4a8c07386940a/testutils/testcluster/testcluster.go#L127)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Make a copy of `args.ServerArgs` instead of mutating the value that was passed in, so that the same ClusterArgs can be used more than once. (this is true for all of the mutations you make to args.ServerArgs, but especially this one since it is the one that varies per server)
</details>
the `ClusterArgs` is taken by this function by value, and the `ServerArgs` inside of it is a value, so these are all private changes

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 18 files at r11, 8 of 9 files at r12, 7 of 9 files at r13.
Review status: 18 of 20 files reviewed at latest revision, 28 unresolved discussions, some commit checks pending.

---

*[server/testserver.go, line 384 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KM_fyQaHjsWBc4poo7n:-KMeF4Tqo6vyC6AAd-wG:-1669082919) ([raw file](https://github.com/cockroachdb/cockroach/blob/d33285fa6fcc1eb26d17a891bed4a8c07386940a/server/testserver.go#L384)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> What's the disadvantage of accessors like this? Making a field (not even a method) of a struct as big and important as `Server` public seems like a big hammer. People are gonna start using them. With accessors like this, at least public access is contained to tests and we can trace what tests are doing involved things when we need to refactor.
> Maybe an alternative you'd find more palatable would be to move these accessors into a struct under `Server`, e.g. `server.TestAccess`? I've opted against that initially because I wasn't convinced that giving access like this to tests is even the right thing to do, maybe we'll come up with something more targeted for `TestCluster`'s usage in the future.
</details>
FWIW, I'm OK with the accessor.

---

*[storage/store.go, line 2577 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448#-KMURHfaQIL4l_sIWU-d:-KMeEjFsnmI1f_gik5re:75178649) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4e7deb279dfd0589ed97c63b1f8e39fd6d946b0/storage/store.go#L2577)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> well the methods are just for tests in the first place, so I don't really see a benefit in having a public version in `helpers_test`. I guess we could append a `ForTest` to all of their names, but I don't think it's worth it.
> I'll add a communal comment before their definition.
</details>
Huh? "public version in `helpers_test`" is an oxymoron.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7448)*
<!-- Sent from Reviewable.io -->
